### ISSUE TITLE
Writing to file vs writing to stdout - or, working with `buf`

### ISSUE THREAD

**Issue by nick-kentik** [joeycumines/protoc-gen-go-copy#1]
I came across this library while trying to resolve some `copylocks` lints - It seems perfect for what I'm trying to do. However, trying to use it as part of [`buf generate`](https://docs.buf.build/generate/usage), with a template like so:

```yaml
version: v1beta1
plugins:
  - name: go-copy
    out: gen/go
    opt:
      - paths=source_relative
```

The run fails with a flood of messages like this:

```
open grpc/health/v1/health_copy.pb.go: no such file or directory
```

`buf` executes the plugin directly, rather than going through `protoc`, and the directory it gives to the plugin doesn't actually exist. So this call fails: https://github.com/joeycumines/protoc-gen-go-copy/blob/master/main.go#L74

Reading through https://rotemtam.com/2021/03/22/creating-a-protoc-plugin-to-gen-go-code/ , it does seem that having a plugin write to the filesystem is unexpected - instead we "should" be calling `f := gen.NewGeneratedFile(filename, file.GoImportPath)` and then `f.P(...)` to write the file to stdout, rather than to disk. `protoc` or `buf` then writes the file to disk on our behalf.

Is this a change you'd be willing to make? I experimented with this minimal change:

<details>

```diff
diff --git a/main.go b/main.go
index 08afaaa..2b6c873 100644
--- a/main.go
+++ b/main.go
@@ -190,7 +190,6 @@ generated for the receiver type, with a name starting with Get.`, x.ShallowCopyM
                }
 
                if err := func() error {
-                       var reader io.Reader
                        if write {
                                r, w := io.Pipe()
                                defer r.Close()
@@ -206,9 +205,12 @@ generated for the receiver type, with a name starting with Get.`, x.ShallowCopyM
                                        }
                                        _ = w.CloseWithError(err)
                                }()
-                               reader = r
+                               g := x.Plugin.NewGeneratedFile(filename, file.GoImportPath)
+                               _, err := io.Copy(g, r)
+                               return err
                        }
-                       return x.Write(filename, reader)
+
+                       return nil
                }(); err != nil {
                        return err
                }
```

</details>

This made `buf` Just Work[tm], and I got a flood of generated files, written to where I expected them to be. An example:

<details>

```go
// Code generated by protoc-gen-go-copy. DO NOT EDIT.
// source: grpc/status/status.proto

package status

import "google.golang.org/protobuf/types/known/anypb"

// Proto_ShallowCopy copies fields, from v to the receiver, using field getters.
// Note that v is of an arbitrary type, which may implement any number of the
// field getters, which are defined as any methods of the same signature as those
// generated for the receiver type, with a name starting with Get.
func (x *Status) Proto_ShallowCopy(v interface{}) {
	switch v := v.(type) {
	case *Status:
		x.Code = v.GetCode()
		x.Message = v.GetMessage()
		x.Details = v.GetDetails()
	default:
		if v, ok := v.(interface{ GetCode() rune }); ok {
			x.Code = v.GetCode()
		}
		if v, ok := v.(interface{ GetMessage() string }); ok {
			x.Message = v.GetMessage()
		}
		if v, ok := v.(interface{ GetDetails() []*anypb.Any }); ok {
			x.Details = v.GetDetails()
		}
	}
}

// Proto_ShallowClone returns a shallow copy of the receiver or nil if it's nil.
func (x *Status) Proto_ShallowClone() (c *Status) {
	if x != nil {
		c = new(Status)
		c.Code = x.Code
		c.Message = x.Message
		c.Details = x.Details
	}
	return
}
```

</details>

**Comment by joeycumines** [comment]
Hey @nick-kentik, cheers for the well written issue, I'll take a closer look at this over the weekend.

What you're saying sounds reasonable. I haven't used this plugin with `buf` before, so I'm not surprised you ran into an issue.

It has been a while and I've forgotten why I wrote the file directly - it's plausible that there's a reason I'll need to retain the existing behavior. That said, even if that's the case, it seems an option to switch the behavior would resolve your issue. I'll investigate.

**Comment by nick-kentik** [comment]
Great, thanks for looking! This would have been a PR but I couldn't understand the nil `reader` thing :sweat_smile:

**Comment by joeycumines** [comment]
It appears to have been a simple fix @nick-kentik, basically just your proposed solution.

For whatever reason, I had decided I didn't want it generating files in cases where there was no content (avoiding empty source files). The nil reader thing was just to handle deleting files, as part of that behavior, since otherwise it might leave the old file there, in certain edge cases. Strange choice, in hindsight.

I've killed off the old behavior, replacing it with the protogen package's file writing mechanism.
